Try to add proxy-judge feature

OmeletWithoutEgg opened the pull request:
An attempt using ActiveJob to add proxy-judge feature for TIOJ.

TODOs
- choose language according to submission.compiler.name
- make sure active jobs are well-maintained
- rejudge functionality
- add more files in judges/
- add information about proxy judge in problem edit panel

OmeletWithoutEgg committed:
Add basic proxy judge feature

OmeletWithoutEgg committed:
Add parameters for proxy judge

OmeletWithoutEgg committed:
Add another type of proxy judge

adrien1018 commented:
Nice! I will take a closer look and do some testing later. A few things I already noticed and would like to change:

- Separate proxy judge option to a boolean field instead of using a special judge option, or just use the `proxy_judge_type` field (which will give a more natural UI, for example selecting proxy judge will hide all native judge settings in the Judge Settings tab and show the proxy judge settings)
    - `proxy_judge_type` can be an enum field
- Reuse the `queued`, `received` and `Validating` verdict

Also it would probably be a good idea to change the Compiler table to store the supported compilers for each OJ (though I am still not entirely sure how to integrate this change with current features such as contest compiler banned list or submission search).

OmeletWithoutEgg committed:
Change proxy judge type to enum and rewrite related logics

OmeletWithoutEgg committed:
Update proxy judge job

- Add different comment style for all languages
- Return JE when exception
- Try to add broadcasting

OmeletWithoutEgg committed:
Fix missing bracket in problem edit

OmeletWithoutEgg committed:
Add different languages to proxy judge file

OmeletWithoutEgg committed:
Add basic description in problem form

OmeletWithoutEgg commented:
I've made some changes, including rewriting the proxy judge type as an enum.
Also, the compiler supported by each judge is now accessible in `PROXY_COMPILERS` hashmap. I'm not sure if there's a better approach.

About the UI layout of the problem form, I'm not sure where should the proxy judge option be placed and what should be hidden when the proxy judge option is selected.

adrien1018 committed:
Decouple proxy judge submit and fetch result

adrien1018 committed:
Revise problem settings UI

adrien1018 commented:
I would like to decouple submission and result fetching, because this way it can tolerate server restart.
I'm still researching for a clean way to run multiple jobs at the same time and also avoid races (i.e. multiple jobs trying to fetch the same submission at the same time), but currently Resque's queue mechanism seems promising.

I did some rewrite to prepare for decoupling (storing proxy judge information in submission table), and also fixed the POJ submission identification issue.

adrien1018 committed:
Fix collation in migration

OmeletWithoutEgg commented on db/migrate/20240624043423_add_proxy_args_to_submission.rb:
Should we use `proxy_judge_type` or `proxyjudge_type`? I switched to `proxyjudge` since there is `specjudge` and I think underscore in the CSS class name in the problem edit form would be a little strange.

adrien1018 replied:
Fixed.

adrien1018 committed:
Rename to proxyjudge

OmeletWithoutEgg commented on app/jobs/judges/CF.rb:
```diff
@@ -80,7 +72,25 @@ def fetch_submission_status
+    return nil if verdict.nil? || verdict.empty? || verdict.include?("ing")
+    {
+      verdict: format_verdict(verdict),
+      time: verdict.to_i,
```
Is it intended to not give the time and memory here?

adrien1018 replied:
No, just a typo :)

OmeletWithoutEgg committed:
Complete the code of decoupling proxyjudge jobs

adrien1018 committed:
Use fetch_results inferface

adrien1018 committed:
Rename CF to codeforces

adrien1018 committed:
Make proxyjudge work

adrien1018 committed:
Disable edit & rejudge in proxy judge problems

OmeletWithoutEgg committed:
Add new proxy judge type QOJ

OmeletWithoutEgg commented:
At this moment Codeforces and POJ are not available for our proxy judge, so I added another proxy judge type QOJ.

Also, currently if the user chose the wrong compiler it will be JE, and I think it would be better to ban these compiler when making new submission. I'm not sure where is the best place to put a helper to select the class based on `proxyjudge_type`, or there are other approaches to do this.

adrien1018 commented:
I think it works for now. In the future we could probably put this into the compilers DB table, and display the correct compiler information to the users.

Also what did you mean by "Codeforces and POJ are not available"?

OmeletWithoutEgg committed:
Set ban_compiler when updating problems with proxyjudge

OmeletWithoutEgg committed:
Use nonce to identify QOJ proxy-judge submissions

OmeletWithoutEgg committed:
Try to fix encoding error when using mechanize

OmeletWithoutEgg committed:
Fix tioj-proxy-nonce exposing

OmeletWithoutEgg committed:
Merge branch 'main' into features/proxy-judge
